lib/improver/ensemble_calibration/ensemble_calibration_utilities.py

```diff
@@ -95,3 +96,145 @@ def check_predictor_of_mean_flag(predictor_of_mean_flag):
+                   "truth has found nothing matching the metadata information "
+                   "supplied: {}".format(input_dict))
+            raise ValueError(msg)
+        return cubelist
+
+    @staticmethod
+    def _filter_non_matching_cubes(historic_forecasts, truths):
+        """
+        Provide filtering for the historic forecast and truth to make sure
+        that these contain matching validity times. This ensures that any
+        mismatch between the historic forecasts and truth is dealt with.
+
+        Args:
+            historic_forecasts (iris.cube.CubeList):
+                CubeList of historic forecasts that potentially contains
+                a mismatch compared to the truth.
+            truths (iris.cube.CubeList):
+                CubeList of truth that potentially contains a mismatch
+                compared to the historic forecasts.
+
+        Returns:
+            (tuple): tuple containing
+                matching_historic_forecasts (iris.cube.CubeList):
+                    CubeList of historic forecasts where any mismatches with
+                    the truth cubelist have been removed.
+                matching_truths (iris.cube.CubeList):
+                    CubeList of truths where any mismatches with
+                    the historic_forecasts cubelist have been removed.
+
+        """
+        matching_historic_forecasts = iris.cube.CubeList([])
+        matching_truths = iris.cube.CubeList([])
+        for cube in historic_forecasts:
+            coord_values = {"time": iris_time_to_datetime(cube.coord("time"))}
+            constr = iris.Constraint(coord_values=coord_values)
+            truth = truths.extract(constr)
+            if truth:
+                matching_historic_forecasts.append(cube)
+                matching_truths.extend(truth)
+        return matching_historic_forecasts, matching_truths
+
+    def process(self, cubes):
+        """
+        Separate the input cubes into the historic_forecasts and truth based
+        on the metadata information supplied within the input dictionaries.
+
+        Args:
+             cubes (iris.cube.CubeList):
+                CubeList of input cubes that are expected to contain a mixture
+                of historic forecasts and truth.
+
+        Returns:
+            (tuple): tuple containing
+
+                iris.cube.Cube:
+                    A cube containing the historic forecasts.
+                iris.cube.Cube:
+                    A cube containing the truth datasets.
+        """
+        historic_forecasts = self._find_required_cubes_using_metadata(
+            cubes, self.historic_forecast_dict)
+        truths = self._find_required_cubes_using_metadata(
+            cubes, self.truth_dict)
+
+        if len(historic_forecasts) != len(truths):
+            historic_forecasts, truths = (
+                self._filter_non_matching_cubes(historic_forecasts, truths))
```

Comment:
Should we apply this filtering even if the two are the same length? That would avoid any hidden mismatching, e.g.:

I have 15 days worth of data looking at T+10. I am missing the T+10 forecast from day 1, and I am missing the truth on day 5. I still get 14 historic forecasts and 14 truths, they are just mismatched.


In the same vein, do we need to sort these lists? Is there some matching of validity times in the ensemble calibration coefficient calculation to ensure we are looking at equivalent validity times between forecast and truth (I assume so, but could not find it at first glance)?

Reply:
After a conversation with @rust, I've actually moved the filtering function out of this plugin and into the `EstimateCoefficientsForEnsembleCalibration` plugin. This is because as you've noted, there isn't otherwise, any checking for equivalent validity times between the forecast and truth, however, this checking is now provided by the `_filter_non_matching_cubes` function.